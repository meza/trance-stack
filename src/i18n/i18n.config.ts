import type { InitOptions } from 'i18next';

export const supportedLngs = ['en'];
export const defaultNS = ['translation'];
export const fallbackLng = 'en';

const baseConfig: InitOptions = {
  supportedLngs: supportedLngs,
  fallbackLng: fallbackLng,
  defaultNS: defaultNS
};

export const getBaseClientConfig = (): InitOptions => ({
  supportedLngs: baseConfig.supportedLngs,
  fallbackLng: baseConfig.fallbackLng,
  defaultNS: baseConfig.defaultNS,
  interpolation: { escapeValue: false },
  react: { useSuspense: false },
  detection: {
    order: ['htmlTag'],
    caches: []
  }
});

export default baseConfig;
