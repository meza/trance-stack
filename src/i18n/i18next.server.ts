import { resolve } from 'node:path';
import { createInstance } from 'i18next';
import Backend from 'i18next-fs-backend';
import { initReactI18next } from 'react-i18next';
import baseConfig from './i18n.config';
import remixI18next from './remix.server';
import type { EntryContext } from '@remix-run/node';
import type { i18n } from 'i18next';

let i18nextInstance: i18n;

const init = async (
  locale: string,
  remixContext: EntryContext
): Promise<i18n> => {
  if (!i18nextInstance) {
    i18nextInstance = createInstance().use(initReactI18next).use(Backend);
    await i18nextInstance.init({
      ...baseConfig,
      lng: locale,
      ns: remixI18next.getRouteNamespaces(remixContext),
      backend: {
        loadPath: resolve('../public/locales/{{lng}}/{{ns}}.json')
      }
    });
  }
  return i18nextInstance;
};

export default init;
