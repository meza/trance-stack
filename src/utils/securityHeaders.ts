import { contentSecurityPolicy } from '~/utils/contentSecurityPolicy';

// https://owasp.org/www-project-secure-headers/ci/headers_remove.json
const headersToClear = {
  'last_update_utc': '2023-02-05 19:51:17',
  'headers': [
    'Host-Header',
    'K-Proxy-Request',
    'Liferay-Portal',
    'Pega-Host',
    'Powered-By',
    'Product',
    'Server',
    'SourceMap',
    'X-AspNet-Version',
    'X-AspNetMvc-Version',
    'X-Atmosphere-error',
    'X-Atmosphere-first-request',
    'X-Atmosphere-tracking-id',
    'X-B3-ParentSpanId',
    'X-B3-Sampled',
    'X-B3-SpanId',
    'X-B3-TraceId',
    'X-CF-Powered-By',
    'X-CMS',
    'X-Content-Encoded-By',
    'X-Envoy-Attempt-Count',
    'X-Envoy-External-Address',
    'X-Envoy-Internal',
    'X-Envoy-Original-Dst-Host',
    'X-Envoy-Upstream-Service-Time',
    'X-Framework',
    'X-Generated-By',
    'X-Generator',
    'X-Mod-Pagespeed',
    'X-Page-Speed',
    'X-Php-Version',
    'X-Powered-By',
    'X-Powered-By-Plesk',
    'X-Powered-CMS',
    'X-Redirect-By',
    'X-Server-Powered-By',
    'X-SourceFiles',
    'X-SourceMap',
    'X-Turbo-Charged-By',
    'X-Varnish-Backend',
    'X-Varnish-Server'
  ]
};

export const securityHeaders = (isDev: boolean): Headers => {
  const owaspHeaders = new Headers();

  owaspHeaders.set('Content-Security-Policy', contentSecurityPolicy(isDev));
  owaspHeaders.set('Cross-Origin-Embedder-Policy', 'require-corp');
  owaspHeaders.set('Cross-Origin-Opener-Policy', 'same-origin');
  owaspHeaders.set('Cross-Origin-Resource-Policy', 'same-origin');
  owaspHeaders.set('Origin-Agent-Cluster', '?1');
  owaspHeaders.set('Permissions-Policy', 'accelerometer=(), camera=(), geolocation=(), gyroscope=(), magnetometer=(), microphone=(), payment=(), usb=()');
  owaspHeaders.set('Referrer-Policy', 'origin-when-cross-origin');
  owaspHeaders.set('Strict-Transport-Security', 'max-age=15552000; includeSubDomains');
  owaspHeaders.set('X-Content-Type-Options', 'nosniff');
  owaspHeaders.set('X-DNS-Prefetch-Control', 'off');
  owaspHeaders.set('X-Download-Options', 'noopen');
  owaspHeaders.set('X-Frame-Options', 'SAMEORIGIN');
  owaspHeaders.set('X-Permitted-Cross-Domain-Policies', 'none');
  owaspHeaders.set('X-XSS-Protection', '0');

  return owaspHeaders;
};

export const addSecurityHeaders = (headers: Headers, isDev: boolean): Headers => {
  const owaspHeaders = securityHeaders(isDev);

  owaspHeaders.forEach((value, key) => {
    headers.set(key, value);
  });

  return headers;
};

export const sanitizeHeaders = (headers: Headers): Headers => {
  headersToClear.headers.forEach((header) => {
    headers.delete(header);
  });

  return headers;
};
