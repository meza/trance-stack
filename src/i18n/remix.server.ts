import { resolve } from 'node:path';
import Backend from 'i18next-fs-backend';
import { RemixI18Next } from 'remix-i18next';
import { supportedLngs, fallbackLng } from './i18n.config';
import type { NewableModule } from 'i18next';

const i18next = new RemixI18Next({
  detection: {
    supportedLanguages: supportedLngs,
    fallbackLanguage: fallbackLng
  },
  i18next: {
    backend: {
      loadPath: resolve('../public/locales/{{lng}}/{{ns}}.json')
    }
  },
  backend: Backend as NewableModule<Backend>
});

export default i18next;
