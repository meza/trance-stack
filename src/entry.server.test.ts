import crypto from 'node:crypto';
import { renderToString } from 'react-dom/server';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { initServerI18n } from '~/i18n';
import { addSecurityHeaders, sanitizeHeaders } from '~/utils/securityHeaders';
import entry from './entry.server';

vi.mock('~/i18n');
vi.mock('~/session.server');
vi.mock('~/utils/securityHeaders');
vi.mock('@remix-run/react');
vi.mock('react-dom/server');
vi.mock('node:crypto');

describe('entry.server', () => {
  beforeEach(() => {
    vi.resetAllMocks();
    vi.mocked(initServerI18n).mockResolvedValue('mocked initServerI18n' as never);
    vi.mocked(renderToString).mockReturnValue('mocked renderToString markup');
    vi.mocked(crypto.randomBytes).mockReturnValue('mocked nonce' as never);
  });

  it('should return the markup', async () => {
    const request = new Request('https://example.com');
    const responseCode = 201;
    const responseHeaders = new Headers();
    const context = {
      staticHandlerContext: {
        loaderData: {
          root: {
            locale: 'en'
          }
        }
      }
    } as never;

    const actualResponse = await entry(request, responseCode, responseHeaders, context);

    expect(await actualResponse.text()).toMatchInlineSnapshot('"<!DOCTYPE html>mocked renderToString markup"');
    expect(actualResponse.status).toBe(201);
  });

  it('should initialise things correctly', async () => {
    vi.stubEnv('NODE_ENV', 'development');

    const request = new Request('https://example.com');
    const responseCode = 200;
    const responseHeaders = new Headers();
    const context = {
      staticHandlerContext: {
        loaderData: {
          root: {
            locale: 'en'
          }
        }
      }
    } as never;

    await entry(request, responseCode, responseHeaders, context);

    expect(initServerI18n).toHaveBeenCalledWith('en', context);
    expect(addSecurityHeaders).toHaveBeenCalledWith(responseHeaders, 'mocked nonce');
    expect(sanitizeHeaders).toHaveBeenCalledWith(responseHeaders);

  });

  it('should return the correct headers', async () => {
    vi.stubEnv('NODE_ENV', 'production');

    const request = new Request('https://example.com');
    const responseCode = 200;
    const responseHeaders = new Headers();
    const context = {
      staticHandlerContext: {
        loaderData: {
          root: {
            locale: 'en'
          }
        }
      }
    } as never;

    const actualResponse = await entry(request, responseCode, responseHeaders, context);
    expect(addSecurityHeaders).toHaveBeenCalledWith(responseHeaders, 'mocked nonce');
    expect(actualResponse.status).toBe(200);
    expect(Object.fromEntries(actualResponse.headers.entries())).toMatchSnapshot();

  });
});
