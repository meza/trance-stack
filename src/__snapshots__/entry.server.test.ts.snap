// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`entry.server > should return the correct headers 1`] = `
{
  "cache-control": "no-cache, max-age=0, s-maxage=0",
  "content-type": "text/html",
}
`;
